### PR TITLE
Re-enable all examples

### LINKED ISSUE #41: Reenable broken example - identity_as_default_projection
Reenable broken example - identity_as_default_projection (https://github.com/beman-project/exemplar/blob/main/examples/CMakeLists.txt#L4), which was temporary disabled due to an error which only occurred on CI.  The example was disabled to be able fast merge a PR before CppCon 2024.

### PR DESCRIPTION
**Pull request by neatudarius**
Closes #41  - re-enable example with ranges support. It is skipped on compilers with no ranges support or when using C++17 or older.

Log sample:
```
# with ranges support
$ rm -rf build/ && cmake -B build -S . -DCMAKE_CXX_STANDARD=20 && cmake --build build && ctest --test-dir build && cmake --install build --prefix /opt/beman.exemplar
...
-- Looking for __cpp_lib_ranges
-- Looking for __cpp_lib_ranges - found
Examples to be built: identity_direct_usage;identity_as_default_projection
-- Configuring done (2.7s)
-- Generating done (0.0s)
-- Build files have been written to: /home/dariusn/git/Beman/exemplar/build

[ 75%] Building CXX object examples/CMakeFiles/beman.exemplar.examples.identity_direct_usage.dir/identity_direct_usage.cpp.o
[ 83%] Linking CXX executable beman.exemplar.examples.identity_direct_usage
[ 83%] Built target beman.exemplar.examples.identity_direct_usage
[ 91%] Building CXX object examples/CMakeFiles/beman.exemplar.examples.identity_as_default_projection.dir/identity_as_default_projection.cpp.o
...
$ tree build/examples/
build/examples/
├── CMakeFiles
│   ├── CMakeDirectoryInformation.cmake
│   ├── beman.exemplar.examples.identity_as_default_projection.dir
│   │   ├── DependInfo.cmake
│   │   ├── build.make
│   │   ├── cmake_clean.cmake
│   │   ├── compiler_depend.make
│   │   ├── compiler_depend.ts
│   │   ├── depend.make
│   │   ├── flags.make
│   │   ├── identity_as_default_projection.cpp.o
│   │   ├── identity_as_default_projection.cpp.o.d
│   │   ├── link.txt
│   │   └── progress.make
│   ├── beman.exemplar.examples.identity_direct_usage.dir
│   │   ├── DependInfo.cmake
│   │   ├── build.make
│   │   ├── cmake_clean.cmake
│   │   ├── compiler_depend.make
│   │   ├── compiler_depend.ts
│   │   ├── depend.make
│   │   ├── flags.make
│   │   ├── identity_direct_usage.cpp.o
│   │   ├── identity_direct_usage.cpp.o.d
│   │   ├── link.txt
│   │   └── progress.make
│   └── progress.marks
├── CTestTestfile.cmake
├── Makefile
├── beman.exemplar.examples.identity_as_default_projection
├── beman.exemplar.examples.identity_direct_usage
└── cmake_install.cmake




# no ranges support
$ rm -rf build/ && cmake -B build -S . -DCMAKE_CXX_STANDARD=17 && cmake --build build && ctest --test-dir build && cmake --install build --prefix /opt/beman.exemplar
...
-- Looking for __cpp_lib_ranges
-- Looking for __cpp_lib_ranges - not found
CMake Warning at examples/CMakeLists.txt:12 (message):
  Missing range support! Skip: identity_as_default_projection


Examples to be built: identity_direct_usage
-- Configuring done (2.7s)
-- Generating done (0.0s)
-- Build files have been written to: /home/dariusn/git/Beman/exemplar/build
...
[ 90%] Building CXX object examples/CMakeFiles/beman.exemplar.examples.identity_direct_usage.dir/identity_direct_usage.cpp.o
...

$ tree build/examples/
build/examples/
├── CMakeFiles
│   ├── CMakeDirectoryInformation.cmake
│   ├── beman.exemplar.examples.identity_direct_usage.dir
│   │   ├── DependInfo.cmake
│   │   ├── build.make
│   │   ├── cmake_clean.cmake
│   │   ├── compiler_depend.make
│   │   ├── compiler_depend.ts
│   │   ├── depend.make
│   │   ├── flags.make
│   │   ├── identity_direct_usage.cpp.o
│   │   ├── identity_direct_usage.cpp.o.d
│   │   ├── link.txt
│   │   └── progress.make
│   └── progress.marks
├── CTestTestfile.cmake
├── Makefile
├── beman.exemplar.examples.identity_direct_usage
└── cmake_install.cmake
```

### CODE REVIEW

File: cmake/CompilerFeatureTest.cmake [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+include(CheckCXXSourceCompiles)
+
+# Determines if the selected C++ compiler has ranges support.
+# Sets 'result_var' to whether support is detected.
+function(beman_check_range_support result_var)
+    # Check if the C++ standard is at least C++20 or later.
+    if(CMAKE_CXX_STANDARD LESS 20)
+        set(${result_var} FALSE PARENT_SCOPE)
+        return()
+    endif()
```
Comment (camio) [inline cmake/CompilerFeatureTest.cmake:14]: `CMAKE_CXX_STANDARD` should not be used to determine if a compiler has C++20 support as it'll give both false negatives and false positives the way it is defined. You have to use `check_cxx_source_compiles` for specific features.

File: cmake/CompilerFeatureTest.cmake [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    check_cxx_source_compiles(
+        "
+        #include <ranges> // C++20 ranges; note that __cpp_lib_ranges is not defined for all compilers
```
Comment (camio) [inline cmake/CompilerFeatureTest.cmake:18]: What compiler doesn't define it?

BTW, when searching for `__cpp_lib_ranges` using GitHub code search I found [this alternative way to check](https://github.com/SparseBLAS/spblas-reference/blob/70e417eaadd9b12c1ff969e01276a8b69e88f10f/CMakeLists.txt#L45).

**Comment by camio** [comment]
Note the CI build failures.

**Comment by wusatosi** [comment]
> Note the CI build failures.

I can give some insight into the CI failure here. Seems like the failed CI are those that doesn't have a C++ version defined. 

This came up to me during refactoring as it's a weird setup, without setting C++ version CMake fallbacks into a platform defined (old) C++ version. I forgot to bring these up.

Basically they fall back to a version < 17 and failed to compile, this the broken checks.

I don't know why it's broken now instead of earlier.

Given the cmake C++ standard conflict @camio mentioned before, we are prob not putting required minimum version in CMake scripts. Thus we should supply the C++ version for all invocation of CMake. 

Those tests are in some sense erroneous.

**Comment by wusatosi** [comment]
Given we will start to conditionally run tests and examples based on C++ versions, we prob want to extend sanitizer-enabled CI tests to all versions as well. I can file this as a separate PR.

**Comment by wusatosi** [comment]
Any interest in progressing this? I can come update it. @neatudarius

**Comment by neatudarius** [comment]
> Any interest in progressing this? I can come update it. @neatudarius

You can take over. I'm busy with iterator_interface this week.

**Comment by wusatosi** [comment]
Should be working now, see: [Unit: gcc 17 Default](https://github.com/beman-project/exemplar/actions/runs/11825757987/job/32950265213?pr=56#step:5:27), that reports `identity_as_default_projection` is not being build. while [Unit: gcc 20 Default](https://github.com/beman-project/exemplar/actions/runs/11825757987/job/32950266483?pr=56#step:5:26) no longer report this.

File: cmake/CompilerFeatureTest.cmake (hunk 2 above)
Comment (wusatosi) [reply cmake/CompilerFeatureTest.cmake]: I have adopted this in my latest contribution.

**Comment by wusatosi** [comment]
Updated default C++ version in preset to C++ 20 (previously C++17) so that all examples are built by default.

**Comment by wusatosi** [comment]
Should be ready! requesting review from original PR author @neatudarius

**Pull request merged**